contracts/eosio.system/src/finalizer_key.cpp

```diff
@@ -0,0 +1,252 @@
+   void system_contract::replace_key_in_finalizer_policy(const name& finalizer, uint64_t old_id, uint64_t new_id) {
+      std::set<uint64_t> next_finalizer_key_ids {_gstate5.last_finalizer_key_ids};
+      std::vector<eosio::finalizer_authority> finalizer_authorities;
+      finalizer_authorities.reserve(next_finalizer_key_ids.size());
+
+      // replace key ID in last_finalizer_key_ids
+      next_finalizer_key_ids.erase(old_id);
+      next_finalizer_key_ids.insert(new_id);
+
+      for( const auto id : next_finalizer_key_ids ) {
+         auto key = _finalizer_keys.find(id);
+         check( key != _finalizer_keys.end(), "key not found in finalizer_keys table for replace_key_in_finalizer_policy");
+         const auto pk = eosio::decode_bls_public_key_to_g1(key->finalizer_key);
+         finalizer_authorities.emplace_back(
+            eosio::finalizer_authority{
+               .description = key->finalizer.to_string(),
+               .weight      = 1,
+               .public_key  = { pk.begin(), pk.end() }
+            }
+         );
+      }
+
+      set_finalizers(finalizer_authorities, next_finalizer_key_ids);
+   }
+
+   // Action to delete a registered finalizer key
+   void system_contract::delfinkey( const name& finalizer, const std::string& finalizer_key ) {
+      require_auth( finalizer );
+
+      auto producer = _producers.find( finalizer.value );
+      check( producer != _producers.end(), "finalizer is not a registered producer");
+
+      // Check the key is registered
+      const auto pk = eosio::decode_bls_public_key_to_g1(finalizer_key);
+      auto idx = _finalizer_keys.get_index<"byfinkey"_n>();
+      auto hash = eosio::sha256(pk.data(), pk.size());
+      auto fin_key = idx.lower_bound(hash);
+      check(fin_key != idx.end(), "finalizer_key was not registered");
+
+      // Check the key belongs to finalizer
+      check(fin_key->finalizer == finalizer, "finalizer_key was not registered by the finalizer");
+      
+      // Cross check finalizers table
+      auto fin = _finalizers.find(finalizer.value);
+      check( fin != _finalizers.end(), "finalizer is not in the finalizers table" );
+      check( fin->num_registered_keys > 0, "num_registered_keys of the finalizer must be greater than one" );
+      
+      if( fin_key->id == fin->active_key_id ) {
+         check( fin->num_registered_keys == 1, "cannot delete an active key unless it is the last registered finalizer key");
+      }
+
+      // Remove the key from finalizer_keys table
+      idx.erase( fin_key );
+
+      // Update finalizers table
+      if( fin->num_registered_keys == 1 ) {
+         // The finalizer does not have any registered keys. Remove it from finalizers table.
+         _finalizers.erase( fin );
+      } else {
+         // Decrement num_registered_keys finalizers table
+         _finalizers.modify( fin, same_payer, [&]( auto& f ) {
+            --f.num_registered_keys;
+         });
+      }
```

Comment:
We need to `set_finalizers` if it was active right? Also need to chose a new finalizer to replace it.

Reply:
@arhag said we don't need take short path when a key is deleted: https://github.com/AntelopeIO/reference-contracts/pull/60#discussion_r1540371632